src/docfx/lib/schema/JsonSchemaTransformer.cs

```diff
@@ -63,70 +63,91 @@ public JsonSchemaTransformer(JsonSchema schema)
                             xrefPropertiesGroupByUid[uid].Item2[key] = new Lazy<JToken>(
                             () =>
                             {
-                                // todo: change transform to `return` model instead of `replace` model
-                                var clonedObj = JsonUtility.DeepClone(obj);
-                                Transform(file, context, propertySchema, clonedObj[key], errors);
-                                return clonedObj[key];
+                                return Transform(file, context, propertySchema, value, errors);
                             }, LazyThreadSafetyMode.PublicationOnly);
                         }
                     }
-                    return schema.XrefProperties;
+                    return (schema.XrefProperties, node);
                 }
 
-                return default;
+                return (default, node);
             });
 
             return (errors, xrefPropertiesGroupByUid);
         }
 
-        private void Transform(Document file, Context context, JsonSchema schema, JToken token, List<Error> errors)
+        private JToken Transform(Document file, Context context, JsonSchema schema, JToken token, List<Error> errors)
         {
-            Traverse(schema, token, (subSchema, node) =>
+            return Traverse(schema, token, (subSchema, node) =>
             {
                 if (node.Type == JTokenType.Array || node.Type == JTokenType.Object)
                 {
                     // don't support array/object transform now
-                    return default;
+                    return (default, node);
                 }
 
-                node.Replace(TransformScalar(subSchema, file, context, node as JValue, errors));
-
-                return default;
+                var transformedScalar = TransformScalar(subSchema, file, context, node as JValue, errors);
+                JsonUtility.SetSourceInfo(transformedScalar, JsonUtility.GetSourceInfo(node));
+                return (default, transformedScalar);
             });
         }
 
-        private void Traverse(JsonSchema schema, JToken token, Func<JsonSchema, JToken, string[]> transform)
+        private JToken Traverse(JsonSchema schema, JToken token, Func<JsonSchema, JToken, (string[], JToken)> transform)
         {
             schema = _definitions.GetDefinition(schema);
+            var sourceInfo = JsonUtility.GetSourceInfo(token);
```

Comment:
To make things simpler, the transformed token does not need to preserve source info, there is no downstream consumers that needs source info afterwards. Same thing can be removed in  line 183.

Reply:
I am not sure about that, will xref need that?

Reply:
nop, after schema transform, it is turned into a json object and pass to template

Reply:
@live1206  confirmed it's not used in xref, so delete it